org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/bundle/BundleSourceTest.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) OSGi Alliance (2020). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.junit5.bundle;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.osgi.framework.Bundle;
+import org.osgi.test.assertj.bundle.BundleAssert;
+import org.osgi.test.assertj.dictionary.DictionaryAssert;
+import org.osgi.test.common.annotation.Property;
+
+public class BundleSourceTest {
+
+	@ParameterizedTest
+	@BundleSource(stateFilter = Bundle.ACTIVE)
+	public void testParam(Bundle bundle) throws Exception {
```

Comment:
I am not sure these tests are sufficient. Sure they test that the supplies argument matches the BundleSource but I don't think it tests that the test method is ever called? What if no bundles match the BundleSource? What if multiple bundles match the BundleSource? The tests should verify these things. 

This basic comment applies to both test classes. There must be some proper way to test ArgumentSource meta-annotated annotations. @kriegfrj ?

Reply:
EngineTestKit would be perfect for this:
* A parent `BundleSourceTest` class should be set up that uses EngineTestKit to actually run Jupiter over supplied test classes (I like to call them "fodder" classes but I'm open to alternative terminology :smile:).
** If there are lots of tests that you would like to run, you can group them into multiple parent classes, eg `BundleSource_SanityCheckingTests` and `BundleSource_ArgumentTests`. See the service tests for an example of this naming convention (I borrowed this naming convention from AssertJ).
* The current test class should be converted to a "fodder" class, which isn't run directly by the tester. This can be done by making them static, package-private inner classes of the parent class. (Although by default Jupiter will run all test classes that are package private, the osgi-test test bundles are set up so that only *public* test classes are included in the `Test-Cases` header, so in this way the tester will skip them.)
* Add a static collection to gather all the bundles on each iteration. Make sure that this is cleared in a `beforeAll()`. The parent test method can then assert on the result to make sure it contains the expected bundles after it has been invoked via the `EngineTestKit`.
* For sanity checking (ie, when you're expecting Jupiter to throw an exception due to some misconfiguration), use the `assertThatTest()` utility method - it will check that an exception is thrown and return a `ThrowableAssert` that you can do your own specific assertions on (eg, assert that the correct message is included in the assertion).

For some examples of this pattern, look at:
https://github.com/osgi/osgi-test/blob/main/org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceExtension_ServiceAwareTest.java
https://github.com/osgi/osgi-test/blob/main/org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceExtension_SanityCheckingTest.java

I really should write some of this up and put it in a Wiki page... I've avoided doing so until now because it has been somewhat experimental, but I think the pattern above is starting to reach the point where it has withstood the test of time.

Reply:
If the above doesn't make sense to @stbischof, I'm happy for this to get merged so that I can work on fleshing out the tests. We're not planning to release 0.11 imminently so I think that's ok?

Reply:
Yeah, that would be fine. @stbischof, if you can address the other comments, we can merge so that this comment can then be worked by @kriegfrj .

Reply:
Just to be clear: my preference is for @stbischof to have a go at what I wrote first and then review it (I think it would be good to spread out that knowledge rather than have me being the only one who knows how to do it); but if he can't make any sense out of it I'll try to pick it up from wherever he leaves off.

Reply:
@kriegfrj I fixed the Header-Filter. It would help a lot if you can take the EngineTestKit-part